File: DONGUK/프로그래머스_WEEK 1/키패드 누르기.py
keypad = {'1':(0,0), '2':(0,1), '3':(0,2), 
          '4':(1,0), '5':(1,1), '6':(1,2), 
          '7':(2,0), '8':(2,1) ,'9':(2,2), 
          '*':(3,0), '0':(3,1), '#':(3,2)}

def getDistance(now, target):
    now_x, now_y = keypad[now]
    tar_x, tar_y = keypad[target]
    
    return abs(tar_x - now_x) + abs(tar_y - now_y)

def solution(numbers, hand):
    answer = ''
    
    now_left, now_right = '*', '#'
    for num in numbers:
        num = str(num)
        
        if num == '1' or num == '4' or num == '7':
            answer += 'L'
            now_left = num
        
        elif num == '3' or num == '6' or num == '9':
            answer += 'R'
            now_right = num
        
        else:
            left_dis = getDistance(now_left, num)
            right_dis = getDistance(now_right, num)

            if left_dis > right_dis:
                answer += 'R'
                now_right = num

            elif left_dis < right_dis:
                answer += 'L'
                now_left = num

            else:
                if hand == "right":
                    answer += 'R'
                    now_right = num

                else:
                    answer += 'L'
                    now_left = num
            
    return answer

File: KIJUNG/PRO_WEEK_02/키패드누르기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int distance(char current, char target){
    int currentX, currentY, targetX, targetY;
    if(current == '*'){
        currentX = 1;
        currentY = 4;
    }
    else if(current == '#'){
        currentX = 3;
        currentY = 4;
    }
    else if(current == '0'){
        currentX = 2;
        currentY = 4;
    }
    else{
        currentX = ((current-48)%3 == 0)? 3:(current-48)%3;
        currentY = (current-48)/3+1;
        currentY -= (currentX==3)? 1:0;
    }

    if(target == '0'){
        targetX = 2;
        targetY = 4;
    }
    else{
        targetX = ((target-48)%3 == 0)? 3:(target-48)%3;
        targetY = (target-48)/3+1;
        targetY -= (targetX==3)? 1:0;
    }
    return abs(currentX - targetX) + abs(currentY - targetY);
}

string solution(vector<int> numbers, string hand) {
    string answer = "";

    char left = '*';
    char right = '#';


    for(char n: numbers){
        
        n+=48;
        if(n == '1' || n == '4' || n == '7'){
            left = n;
            answer += "L";
        }
        else if(n == '3' || n == '6' || n == '9'){
            right = n;
            answer += "R";
        }
        else{
            int leftDist = distance(left, n);
            int rightDist = distance(right, n);

            if(leftDist == rightDist){
                if(hand == "left"){
                    left = n;
                    answer += "L";
                }
                else{
                    right = n;
                    answer += "R";
                }
            }
            else if(leftDist > rightDist){
                right = n;
                answer += "R";
            }
            else{
                left = n;
                answer += "L";
            }
        }
    }

    return answer;
}